### PR TITLE
Webpack v3to v4

### PR DESCRIPTION
**Pull request by harpreetkhalsagtbit**
* [X]  Passed [sanity tests](http://bit.ly/sttm-sanity-tests).
* [X]  Ran `npm test` & fixed newly introduced lint errors.
* [X]  Checked console for errors.
informed abt console errs

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -74,8 +74,6 @@
   <!-- Polyfills -->
   <script src="https://cdn.polyfill.io/v2/polyfill.min.js?features=fetch,Object.entries,Array.prototype.includes,String.prototype.includes,Map,Set,requestAnimationFrame,Array.prototype.@@iterator"></script>
 
-  <!-- Vendor Scripts -->
-  <script src="/assets/js/vendor.js"></script>
```
Comment (bogas04) [inline index.html]: Add chunks/vendor.js

```html
  <script src="/assets/js/chunks/vendor.js"></script>
```

File: webpack.config.js [hunk 2]
```diff
@@ -19,38 +19,20 @@ const plugins = PRODUCTION
...
-  'shabados',
-  'redux',
-  'react-redux',
-  'react-router-dom',
-];
-
 const app = path.resolve(__dirname, 'src', 'js', 'index.js');
 
 module.exports = {
+  mode: PRODUCTION || 'development',
```
Comment (bogas04) [inline webpack.config.js]: This won't work, use ternary operator

```javascript
PRODUCTION ? 'production' : 'development'
```

File: webpack.config.js [hunk 3]
```diff
@@ -60,6 +42,10 @@ module.exports = {
     publicPath: 'assets/js/',
   },
   plugins,
+  optimization: {
+    noEmitOnErrors: true, // NoEmitOnErrorsPlugin
+    concatenateModules: true //ModuleConcatenationPlugin
```
Comment (bogas04) [inline webpack.config.js]: Add vendor as a splitChunk

```diff
optimization: {
    noEmitOnErrors: true, // NoEmitOnErrorsPlugin
-    concatenateModules: true //ModuleConcatenationPlugin
+   concatenateModules: true, //ModuleConcatenationPlugin
+   splitChunks: {
+     cacheGroups: {
+       // All dependencies in `node_modules` become part of vendor chunk
+       vendors: {
+         test: /[\\/]node_modules[\\/]/,
+         name: 'vendor',
+         chunks: 'all',
+       },
+     },
    },
  },
```

**Comment by bogas04** [comment]
Vahiguru Jee Ka Khalsa, Vahiguru Jee Ki Fateh Jee.

Thank you so much for this PR, it gives a lot of wins to us for free!

## Before

| # | Asset | Size | Chunks | Name
|:--|:--|:--|:--|:--
1|chunks/NotFound.js|  2.04 kB    |   5  [emitted]  |NotFound
2|chunks/Shabad.js|  52.7 kB    |   0  [emitted]  |Shabad
3|chunks/Ang.js|  54.4 kB    |   2  [emitted]  |Ang
4|chunks/Search.js|  33.4 kB    |   3  [emitted]  |Search
5|chunks/TermsOfService.js|  8.87 kB    |   4  [emitted]  |TermsOfService
6|chunks/Hukamnama.js|  53.9 kB    |   1  [emitted]  |Hukamnama
7|chunks/Help.js|  14.3 kB    |   6  [emitted]  |Help
8|chunks/GranthIndex.js|  6.08 kB    |   7  [emitted]  |GranthIndex
9|chunks/About.js|  2.81 kB    |   8  [emitted]  |About
10|app.js|   103 kB    |   9  [emitted]  |app
11|vendor.js|   182 kB    |  10  [emitted]  |vendor

Total (513.5KB)

## After


| # |Asset    |  Size|  Chunks     |        Chunk Names
|:--|:--|:--|:--|:--
1|chunks/GranthIndex.js | 5.92 KiB     |  6  [emitted] |  GranthIndex
2|chunks/Ang\~Hukamnama\~Search\~Shabad.js | 25.6 KiB     |  0  [emitted] | Ang\~Hukamnama\~Search\~Shabad
3|chunks/Shabad.js | 2.05 KiB     |  2  [emitted] | Shabad
4|chunks/Search.js | 6.91 KiB     |  3  [emitted] | Search
5|chunks/Hukamnama.js | 3.19 KiB     |  4  [emitted] | Hukamnama
6|chunks/Help.js | 13.9 KiB     |  5  [emitted] | Help
7|chunks/Ang\~Hukamnama\~Shabad.js | 23.7 KiB     |  1  [emitted]  |Ang\~Hukamnama\~Shabad
8|chunks/Ang.js | 3.69 KiB     |  7  [emitted] | Ang
9|chunks/About.js | 2.73 KiB     |  8  [emitted] | About
10|chunks/TermsOfService.js | 8.64 KiB     |  9  [emitted] | TermsOfService
11|chunks/NotFound.js | 1.97 KiB     | 10  [emitted]  |NotFound
12|chunks/vendor.js |  176 KiB     | 11  [emitted]  |vendor
13|app.js |  103 KiB     | 12  [emitted]  |app

Total (377.3KB)

As we can see, 2 new chunks have been created by Webpack. It was smart enough to figure out that Ang/Hukamnama/Search/Shabad all use "ShareControls" while Ang/Hukamnama/Shabad all use "ShabadContent", and thus created new chunks for those. This is actually efficient. Big props to webpack team for navigating hierarchy to figure all this out. This resulted in overall smaller size of JS as those 4 routes aren't duplicating those heavy components. Though, it doesn't matter much for initial load, as those were dynamically loaded anyway.

For initial boot, we still lost some KBs (285KB to 279KB).

Webpack v4 also promises reduced compilation time, this'll mean faster deployments at jenkins level, and faster development :D

**Comment by bogas04** [comment]
One more thing, we need to prefetch chunks/Ang~Hukamnama~Shabad.js & chunks/Ang~Hukamnama~Search~Shabad.js now.

**Comment by harpreetkhalsagtbit** [comment]
@bogas04 
Updated PR with the new changes

**Pull request merged**